### PR TITLE
RFC: define a single `otk.op.join`

### PR DESCRIPTION
**Pull request by mvo5**
I wonder if we can simply avoid having the user care about the type when doing a join operation. It seems mostly a burden to the user and we need to check during compilation of the otk file anyway so maybe we can just leave it out?

P.S. Having an example use-case for both would be really good, I'm wonder if we only should add things to the spec once we actually use it in the example/* tree ?

### CODE REVIEW

**Comment by supakeen** [comment]
My reasoning for this was that having separate operations per type allows us to have separate arguments per type. For a map erroring on duplicates (or overwriting them) makes sense. For a sequence less so.

**Comment by mvo5** [comment]
> My reasoning for this was that having separate operations per type allows us to have separate arguments per type. For a map erroring on duplicates (or overwriting them) makes sense. For a sequence less so.

Thanks! I'm not sure I follow - if `otk.op.join` would be used on two maps with duplicates it would still error. If it is used on two sequences with duplicates it would not. Mixing the two is not allowed. What am I missing?

**Comment by supakeen** [comment]
> > My reasoning for this was that having separate operations per type allows us to have separate arguments per type. For a map erroring on duplicates (or overwriting them) makes sense. For a sequence less so.
> 
> Thanks! I'm not sure I follow - if `otk.op.join` would be used on two maps with duplicates it would still error. If it is used on two sequences with duplicates it would not. Mixing the two is not allowed. What am I missing?

Wanting to define the behavior, for example:

```
otk.op.map.join:
  duplicates: overwrite
  values:
    - key:
        one
    - key:
        two
```

The `duplicates` key would not be very sensical on the `otk.op.seq.join` I feel?

**Comment by mvo5** [comment]
> > > My reasoning for this was that having separate operations per type allows us to have separate arguments per type. For a map erroring on duplicates (or overwriting them) makes sense. For a sequence less so.
> > 
> > 
> > Thanks! I'm not sure I follow - if `otk.op.join` would be used on two maps with duplicates it would still error. If it is used on two sequences with duplicates it would not. Mixing the two is not allowed. What am I missing?
> 
> Wanting to define the behavior, for example (don't pin me on this one ;)).
> 
> ```
> otk.op.map.join:
>   duplicates: overwrite
>   values:
>     - key:
>         one
>     - key:
>         two
> ```
> 
> The `duplicates` key would not do anything on the `otk.op.seq.join` I feel?

Thank you! That is interesting (and the context I was looking for!). We could start with what popular languages like python are doing:
1. duplicated keys in a map get replaced by "later" joins (essentially meaning overwrite)
2. seq are just joined

I like the idea of the users having a say here though, so once the above is not good enough anymore, we could add rules like 
```
duplicates: [error, overwrite]
```
and then we could decide to error if "overwrite" is used on sequences. Or we could also "bend" the definition and "de-duplicate/replace/overwrite" keys in a seqence if later sequences also define it (not sure I like it or not but it's an option :)

**Comment by mvo5** [comment]
I fixed the conflicts, sorry that this took me a bit

File: doc/directives.md [hunk 1]
```diff
@@ -134,10 +131,15 @@ otk.define:
   b:
     b: 2
   c:
-    otk.op.hash.merge:
+    otk.op.merge:
```
[the remark this answers is not in the document]
Comment (mvo5) [reply doc/directives.md]: Thanks, updated. Sorry that it took so long

**Comment by supakeen** [comment]
This does still need the implementation change at: https://github.com/osbuild/otk/blob/main/src/otk/directive.py#L107-L154 and https://github.com/osbuild/otk/blob/main/src/otk/transform.py#L51-L52 :)

**Review by kingsleyzissou — COMMENTED** [review]
> Duplicate keys in maps are considered an error.

It might be nice to have a test for this (unless there is one already and I missed it)

File: src/otk/directive.py [hunk 2]
```diff
@@ -96,53 +96,50 @@ def op(ctx: Context, tree: Any, key: str) -> Any:
...
         )
 
     return list(itertools.chain.from_iterable(values))
 
 
-@tree.must_be(dict)
-@tree.must_pass(tree.has_keys(["values"]))
-def op_map_merge(ctx: Context, tree: dict[str, Any]) -> Any:
-    """Merge two dictionaries. Keys from the second dictionary overwrite keys
+def _op_map_join(ctx: Context, values: List[dict]) -> Any:
```
Comment (supakeen) [inline src/otk/directive.py:136]: `List` wasn't imported from `typing`.

**Comment by supakeen** [comment]
> > Duplicate keys in maps are considered an error.
> 
> It might be nice to have a test for this (unless there is one already and I missed it)

There's no test for it (yet) and the code doesn't *actually* check this: https://github.com/osbuild/otk/blob/9ead440ab1282dec1e25877180d105b38fbef177/src/otk/directive.py#L148

It should straightforward to add, or reconsider if it should be an error.

**Comment by supakeen** [comment]
@mvo5 you imported `List` in the wrong module ;)

**Comment by mvo5** [comment]
> @mvo5 you imported `List` in the wrong module ;)

Silly me, sorry for that! Fwiw, `make type` gives me a bunch of errors so I did not see this one but it's not really important at this point (the wrong List import is hopefully fixed now).

**Review by supakeen — APPROVED** [review]

**Pull request merged**